crates/derive/src/sources/calldata.rs

```diff
@@ -0,0 +1,93 @@
+//! CallData Source
+
+use crate::traits::{AsyncIterator, ChainProvider};
+use crate::types::BlockInfo;
+use crate::types::StageError;
+use crate::types::StageResult;
+use alloc::boxed::Box;
+use alloc::vec::Vec;
+use alloy_primitives::{Address, Bytes};
+use async_trait::async_trait;
+
+/// A data iterator that reads from calldata.
+#[derive(Debug, Clone)]
+#[allow(dead_code)]
+pub struct CalldataSource<CP>
+where
+    CP: ChainProvider + Send,
+{
+    /// The chain provider to use for the calldata source.
+    chain_provider: CP,
+    /// The address of the batcher contract.
+    batcher_address: Address,
+    /// Block Ref
+    block_ref: BlockInfo,
+    /// The L1 Signer.
+    signer: Address,
+    /// Current calldata.
+    calldata: Vec<Bytes>,
+    /// Whether the calldata source is open.
+    open: bool,
+}
+
+impl<CP: ChainProvider + Send> CalldataSource<CP> {
+    /// Creates a new calldata source.
+    pub fn new(
+        chain_provider: CP,
+        batcher_address: Address,
+        block_ref: BlockInfo,
+        signer: Address,
+    ) -> Self {
+        Self {
+            chain_provider,
+            batcher_address,
+            block_ref,
+            signer,
+            calldata: Vec::new(),
+            open: false,
+        }
+    }
+
+    /// Loads the calldata into the source if it is not open.
+    async fn load_calldata(&mut self) -> anyhow::Result<()> {
+        if self.open {
+            return Ok(());
+        }
+
+        let info = self
+            .chain_provider
+            .block_info_and_transactions_by_hash(self.block_ref.hash)
+            .await?;
+
+        self.calldata = info
+            .1
+            .iter()
+            .filter_map(|tx| {
+                if tx.to() != Some(self.batcher_address) {
+                    return None;
+                }
+                if tx.from() != Some(self.signer) {
+                    return None;
+                }
+                Some(tx.data())
+            })
+            .collect::<Vec<_>>();
+
+        self.open = true;
+
+        Ok(())
+    }
+}
+
+#[async_trait]
+impl<CP: ChainProvider + Send> AsyncIterator for CalldataSource<CP> {
+    type Item = StageResult<Bytes>;
+
+    async fn next(&mut self) -> Option<Self::Item> {
+        if self.load_calldata().await.is_err() {
+            return Some(Err(StageError::BlockFetch(self.block_ref.hash)));
+        }
+
+        Some(self.calldata.pop().ok_or(StageError::Eof))
```

Comment:
This is incorrect. The `pop()` function removes the last element from the iterator, we want to `pop_front`. Maybe `VecDequeue` is helpful here?

ref: https://github.com/ethereum-optimism/optimism/blob/develop/op-node/rollup/derive/calldata_source.go#L67-L73

Reply:
Since `VecDeque` is not available in the alloc crate (?), I've modified the `self.calldata` assignment to iterate in reverse. This allows us to just `.pop()` :)

Reply:
I think it should be available @ [`alloc::collections::VecDeque`](https://doc.rust-lang.org/stable/alloc/collections/index.html#reexport.VecDeque) - we also use it in the `ChannelBank` + `FrameQueue` atm :D